Library/Homebrew/utils/curl.rb

```diff
@@ -359,6 +359,7 @@ def curl_http_content_headers_and_checksum(
 
       if status.success?
         file_contents = File.read(file.path)
+        file_contents.encode!(Encoding::UTF_8, invalid: :replace) if headers["content-type"]&.start_with?("text/")
```

Comment:
My only concern here is handling binary data. But I don't really know if forced encoding could break something for our use cases.

Reply:
Yeah `invalid: :replace` is a dangerous for binary files.

The correct way to handle this IMO would be:

* If `charset=` exists in the Content-Type HTTP header, pass that to `Encoding.find` and check if it's valid (i.e. no `ArgumentError`). If so the charset should be passed to `File.read` via the `encoding:` kwarg.
* To be comprehensive, we ideally should also parse for `<meta>` tags, but it's probably a bit overkill.
* Instead of deleting invalid characters here, we maybe should instead scope this to the `gsub` calls to be safer. There is also a `String#scrub` method you can use rather than `String#encode!`.

Reply:
> Yeah `invalid: :replace` is dangerous for binary files.

Good point, 👍🏻 to not trying to change them 😅

Reply:
Thanks for the feedback; I'll revise the PR!

Reply:
> If charset= exists in the Content-Type HTTP header, pass that to Encoding.find and check if it's valid (i.e. no ArgumentError). If so the charset should be passed to File.read via the encoding: kwarg.

Added

> Instead of deleting invalid characters here, we maybe should instead scope this to the gsub calls to be safer (since we must already be dealing with text data in those cases). There is also a String#scrub method you can use rather than String#encode!. This is already used in livecheck.rb.

Changed